File: hackerearth-vignan/III-SEM/challenge_13/Q1_infixToPrefix.py
def precedence(op):
    return {'+': 1, '-': 1, '*': 2, '/': 2}.get(op, 0)

def infix_to_prefix(infix):
    infix = infix[::-1].translate(str.maketrans('()', ')('))
    operators, prefix = [], ""

    for c in infix:
        if c.isalnum():
            prefix += c
        elif c == '(':
            operators.append(c)
        elif c == ')':
            prefix += operators.pop() * (operators[-1] != '(')
            operators.pop()  # Remove '('
        elif c in {'+', '-', '*', '/'}:
            while operators and precedence(operators[-1]) >= precedence(c):
                prefix += operators.pop()
            operators.append(c)

    return prefix + ''.join(operators[::-1])

if __name__ == "__main__":
    infix_expression = input("")
    print(infix_to_prefix(infix_expression)[::-1])

File: hackerearth-vignan/III-SEM/challenge_13/Q1_infixToPrefix.cpp
#include <bits/stdc++.h>
using namespace std;

int precedence(char op) {
    if (op == '+' || op == '-')
        return 1;
    if (op == '*' || op == '/')
        return 2;
    return 0;
}

string infixToPrefix(string infix) {
    reverse(infix.begin(), infix.end());
    for (char &c : infix) {
        if (c == '(')
            c = ')';
        else if (c == ')')
            c = '(';
    }
    stack<char> operators;
    string prefix;
    for (char c : infix) {
        if (isalnum(c)) {
            prefix += c;
        } else if (c == '(') {
            operators.push(c);
        } else if (c == ')') {
            while (!operators.empty() && operators.top() != '(') {
                prefix += operators.top();
                operators.pop();
            }
            operators.pop();
        } else if (c == '+' || c == '-' || c == '*' || c == '/') {
            while (!operators.empty() && precedence(operators.top()) >= precedence(c)) {
                prefix += operators.top();
                operators.pop();
            }
            operators.push(c);
        }
    }
    while (!operators.empty()) {
        prefix += operators.top();
        operators.pop();
    }
    reverse(prefix.begin(), prefix.end());
    return prefix;
}

int main() {
    string infix_expression;
    cin >> infix_expression;
    string prefix_expression = infixToPrefix(infix_expression);
    cout << prefix_expression << endl;
    return 0;
}
